YSQL Datamodeling (#22554)
yugabyte/yugabyte-db: docs/content/preview/develop/learn/data-modeling-ysql.md

---
title: YSQL data modeling in YugabyteDB
headerTitle: Distributed SQL Data modeling
linkTitle: Data modeling
description: Learn data modeling in YSQL and how to identify the patterns used to access data and the types of queries to be performed.
menu:
  preview:
    identifier: data-modeling-ysql
    parent: learn
    weight: 100
type: docs
---

{{<api-tabs>}}

Data modeling involves designing the database schema for efficient storage and access. In a distributed SQL database like YugabyteDB, table data is split into tablets and distributed across multiple nodes within the cluster, allowing applications to connect to any node for storing and retrieving data. Reads and writes can span multiple nodes, so it's crucial to consider how table data is partitioned and distributed when modeling your data.

In YugabyteDB, data is stored as rows and columns in tables which are organized under schemas and databases.

{{<lead link="../../../explore/ysql-language-features/databases-schemas-tables">}}

> To understand more about creating and managing tables, schemas and databases, see [Schemas and tables](../../../explore/ysql-language-features/databases-schemas-tables)

Reviewer: ```suggestion
To understand more about creating and managing tables, schemas, and databases, see [Schemas and tables](../../../explore/ysql-language-features/databases-schemas-tables).
```

{{</lead>}}

> When modeling your data for YugabyteDB, you must understand the 2 [data distribution](../../../explore/going-beyond-sql/data-sharding) schemes, Hash and Range sharding in detail. This is very significant for designing your tables and indexes for fast retrieval and storage.

Reviewer: ```suggestion
To design your tables and indexes for fast retrieval and storage in YugabyteDB, you first need to understand the two [data distribution](../../../explore/going-beyond-sql/data-sharding) schemes, Hash and Range sharding, in detail.
```

> In YugabyteDB the sharding and ordering of data in the tables and indexes is governed by the primary key of the table and index respectively. Let us dig deeper into designing the primary keys with some examples.

Reviewer: ```suggestion
In YugabyteDB, the sharding and ordering of data in the tables and indexes is governed by the primary key of the table and index respectively.
```

  (10,'Diane',51,94083,False),    (11,'Ashley',42,94083,False),  (12,'Jacqueline',58,94085,False),
  (13,'Benjamin',49,94084,False), (14,'James',48,94083,False),   (15,'Ann',43,94083,False),
  (16,'Aimee',47,94085,True),     (17,'Michael',49,94085,False), (18,'Rebecca',40,94085,False),
  (19,'Kevin',45,94085,True),     (20,'James',45,94084,False),   (21,'Sandra',60,94085,False),
  (22,'Kathleen',40,94085,True),  (23,'William',42,94084,False), (24,'James',42,94083,False),
  (25,'Tyler',50,94085,False),    (26,'James',49,94085,True),    (27,'Kathleen',55,94083,True),
  (28,'Zachary',55,94083,True),   (29,'Rebecca',41,94085,True),  (30,'Jacqueline',49,94085,False),
  (31,'Diane',48,94083,False),    (32,'Sarah',53,94085,True),    (33,'Rebecca',55,94083,True),
  (34,'William',47,94085,False),  (35,'William',60,94085,True),  (36,'Sarah',53,94085,False),
  (37,'Ashley',47,94084,True),    (38,'Ashley',54,94084,False),  (39,'Benjamin',42,94083,False),
  (40,'Tyler',47,94085,True),     (41,'Michael',42,94084,False), (42,'Diane',50,94084,False),
  (43,'Nancy',51,94085,False),    (44,'Rebecca',56,94085,False), (45,'Tyler',41,94085,True);
```

</details>

{{<note>}}

> To explain the behavior of the queries we will use **explain (analyze, dist, costs off)**. In practice, you do not need this unless you are trying to optimize the performance. For more details, see [Analyze queries](../../../explore/query-1-performance/explain-analyze)

Reviewer: ```suggestion
To explain the behavior of the queries, the examples use **explain (analyze, dist, costs off)**. In practice, you do not need to do this unless you are trying to optimize performance. For more details, see [Analyze queries](../../../explore/query-1-performance/explain-analyze).
```
